src/collapse/xctrace.rs

```diff
@@ -0,0 +1,455 @@
+            FRAME => matches!(self, Self::Frame(_)),
+            other => matches!(self, Self::Other(tag) if tag.as_bytes() == other),
+        }
+    }
+}
+
+#[derive(Default)]
+struct TraceQueryResultState {
+    nodes: Vec<Node>,
+}
+
+#[derive(Default)]
+struct NodeState {
+    rows: Vec<Row>,
+}
+
+#[derive(Default)]
+struct RowState {
+    backtrace: Option<Arc<Backtrace>>,
+}
+
+#[derive(Default)]
+struct BacktraceState {
+    id: u64,
+    frames: Vec<Arc<Frame>>,
+}
+
+#[derive(Default)]
+struct FrameState {
+    id: u64,
+    name: Vec<u8>,
+}
+
+struct Node {
+    rows: Vec<Row>,
+}
+
+struct Row {
+    backtrace: Arc<Backtrace>,
+}
+
+struct Backtrace {
+    id: u64,
+    frames: Vec<Arc<Frame>>,
+}
+
+struct Frame {
+    id: u64,
+    name: Vec<u8>,
+}
+
+impl Backtrace {
+    fn to_folded(&self) -> Vec<u8> {
+        let mut folded = Vec::new();
+        let mut first = Some(());
+        // Because stack frames are arranged from top to bottom in xctrace's
+        // output, here we use `.rev(`.
+        for frame in self.frames.iter().rev() {
+            if first.take().is_none() {
+                folded.push(b';');
+            }
+            let frame_name = String::from_utf8_lossy(&frame.name);
+            let frame_name = fix_partially_demangled_rust_symbol(&frame_name);
+            folded.extend(frame_name.as_bytes().iter().copied());
+        }
+        folded
+    }
+}
+
+/// Unescapes the text in xml exported from xctrace.
+fn unescape_xctrace_text(x: Cow<'_, [u8]>) -> Vec<u8> {
+    // xctrace shouldn't give us invalid xml text here, therefore
+    // we don't expect the error branch being hit:
+    //
+    // `quick_xml::escape::unescape` will error out if the input is not a valid xml text:
+    // https://github.com/tafia/quick-xml/blob/0793d6a8d006cb5dabf66bf2a25ddbf198305b46/src/escape.rs#L253
+    if let Ok(x) = quick_xml::escape::unescape(&String::from_utf8_lossy(&x)) {
+        x.into_owned().into_bytes()
+    } else {
+        x.into_owned()
+    }
+}
+
+fn get_ref_id_from_attributes(attributes: &Attributes) -> io::Result<u64> {
+    let ref_id = attributes
+        .clone()
+        .filter_map(|x| x.ok())
+        .find_map(|x| (x.key.into_inner() == REF).then_some(x.value));
+    let ref_id = match ref_id {
+        Some(x) => x,
+        None => return invalid_data_error!("No ref id found in attributes"),
+    };
+    let ref_id = String::from_utf8_lossy(&ref_id);
+    match ref_id.parse() {
+        Ok(x) => Ok(x),
+        Err(e) => invalid_data_error!("Unrecognized ref id: {}: {:?}", ref_id, e),
+    }
+}
+
+fn get_id_from_attributes(attributes: &Attributes) -> io::Result<u64> {
+    let id = attributes
+        .clone()
+        .filter_map(|x| x.ok())
+        .find_map(|x| (x.key.into_inner() == ID).then_some(x.value));
+    let id = match id {
+        Some(x) => x,
+        None => return invalid_data_error!("No id found in attributes"),
+    };
```

Comment:
I'd probably use let-else here to make this read slightly nicer.
```suggestion
    let Some(id) = id else {
        return invalid_data_error!("No id found in attributes");
    };
```
There are a couple of other places too where this pattern could work nicely.